codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedIntegerGenerator.kt

```diff
@@ -0,0 +1,187 @@
+        constrainedTypeMetadata.render(writer)
+        writer.rust("struct $name(pub(crate) $inner);")
+        if (constrainedTypeVisibility == Visibility.PUBCRATE) {
+            Attribute.AllowUnused.render(writer)
+        }
+        writer.rustTemplate(
+            """
+            impl $name {
+                /// ${rustDocsInnerMethod(inner)}
+                pub fn inner(&self) -> &$inner {
+                    &self.0
+                }
+
+                /// ${rustDocsIntoInnerMethod(inner)}
+                pub fn into_inner(self) -> $inner {
+                    self.0
+                }
+            }
+
+            impl #{ConstrainedTrait} for $name  {
+                type Unconstrained = $inner;
+            }
+
+            impl #{From}<$inner> for #{MaybeConstrained} {
+                fn from(value: $inner) -> Self {
+                    Self::Unconstrained(value)
+                }
+            }
+
+            impl #{From}<$name> for #{MaybeConstrained} {
+                fn from(value: $name) -> Self {
+                    Self::Constrained(value)
+                }
+            }
+
+            impl #{Display} for $name {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                   ${shape.redactIfNecessary(model, "self.0")}.fmt(f)
+                }
+            }
+
+            impl #{TryFrom}<$inner> for $name {
+                type Error = #{ConstraintViolation};
+
+                /// ${rustDocsTryFromMethod(name, inner)}
+                fn try_from(value: $inner) -> Result<Self, Self::Error> {
+                    if $condition {
+                        Ok(Self(value))
+                    } else {
+                        Err(#{ConstraintViolation}::Size(value))
+                    }
+                }
+            }
+
+            impl #{From}<$name> for $inner {
+                fn from(value: $name) -> Self {
+                    value.into_inner()
+                }
+            }
+
+            impl #{AsRef}<$inner> for $name {
+                fn as_ref(&self) -> &$inner {
+                    self.inner()
+                }
+            }
+            """,
+            "ConstrainedTrait" to RuntimeType.ConstrainedTrait(),
+            "ConstraintViolation" to constraintViolation,
+            "MaybeConstrained" to symbol.makeMaybeConstrained(),
+            "Display" to RuntimeType.Display,
+            "From" to RuntimeType.From,
+            "TryFrom" to RuntimeType.TryFrom,
+            "AsRef" to RuntimeType.AsRef,
+        )
+
+        val constraintViolationModuleName = constraintViolation.namespace.split(constraintViolation.namespaceDelimiter).last()
+        writer.withModule(RustModule(constraintViolationModuleName, RustMetadata(visibility = constrainedTypeVisibility))) {
+            rust(
+                """
+                ##[derive(Debug, PartialEq)]
+                pub enum ${constraintViolation.name} {
+                    Size($inner),
```

Comment:
`Range` may be a better name, to match the trait name?

Reply:
I agree, fixed in https://github.com/awslabs/smithy-rs/pull/2005/commits/6e54c5e4ad228be25c6cb27d54766cc00c31f832